### ISSUE TITLE
[core][oom] large scale workloads reporting OOM deadlock

### ISSUE THREAD

**Issue by clarng** [ray-project/ray#32822]
### What happened + What you expected to happen

This is happening to several of our large scale nightly tests  , e.g. https://github.com/ray-project/ray/pull/32361

The issues is that we have a multi-layer DAG with high fanout, and that makes it likely that all tasks running on a node does not share a parent caller. When the node OOMs, the OOM killer will mistaken this as a deadlock and fail the entire workload.

Here we propose a policy that will avoid this issue and meet our requirements around liveness and deadlock detection
policy details:
- we annotate each driver-spawn task with a priority (timestamp),
- tasks spawned will inherit the priority of its caller
- actor's priority is equal to the highest priority task that is running at the time of OOM, else it is based on the time of actor creation (which is higher than the task) 

the main risk to this proposed policy is that it is greedy vs fair - it prioritizes first comer, for example, if we run N trials and the cluster can only satisfy K, the remaining N-K trials will be killed, whereas in the current policy, we will try to run all N, and if it is not possible to run in parallel, it will fail the workload entirely

### Versions / Dependencies

master

### Reproduction script

https://github.com/ray-project/ray/pull/32361

### Issue Severity

None

**Comment by clarng** [comment]
cc @ericl  @stephanie-wang  @scv119

**Comment by ericl** [comment]
This doesn't happen in any realistic user workloads right? The scenario is a bit contrived. Also, raising OOM is totally fine here--- the goal of the OOM killer is just to protect the node as P0. Automatically scheduling memory is just a nice to have.

I'm concerned the complexity of implementation here will outweigh any concrete benefit.

**Comment by clarng** [comment]
The implementation is pretty straightforward, mostly plumbing

This would happen more likely for larger clusters, and with more than 1 layer of tree / dag
